pypsa/contingency.py

```diff
@@ -154,11 +155,152 @@ def network_lpf_contingency(network, snapshots=None, branch_outages=None):
+
+        flow_upper.update({(branch[0],branch[1],b[0],b[1],sn) : [[(1,network.model.passive_branch_p[b[0],b[1],sn]),(sub.BODF[sub._branches.at[b,"_i"],branch_i],network.model.passive_branch_p[branch[0],branch[1],sn])],"<=",sub._fixed_branches.at[b,"s_nom"]] for b in sub._fixed_branches.index for sn in snapshots})
+
+        flow_upper.update({(branch[0],branch[1],b[0],b[1],sn) : [[(1,network.model.passive_branch_p[b[0],b[1],sn]),(sub.BODF[sub._branches.at[b,"_i"],branch_i],network.model.passive_branch_p[branch[0],branch[1],sn]),(-1,network.model.passive_branch_s_nom[b[0],b[1]])],"<=",0] for b in sub._extendable_branches.index for sn in snapshots})
+
+
+        flow_lower.update({(branch[0],branch[1],b[0],b[1],sn) : [[(1,network.model.passive_branch_p[b[0],b[1],sn]),(sub.BODF[sub._branches.at[b,"_i"],branch_i],network.model.passive_branch_p[branch[0],branch[1],sn])],">=",-sub._fixed_branches.at[b,"s_nom"]] for b in sub._fixed_branches.index for sn in snapshots})
+
+        flow_lower.update({(branch[0],branch[1],b[0],b[1],sn) : [[(1,network.model.passive_branch_p[b[0],b[1],sn]),(sub.BODF[sub._branches.at[b,"_i"],branch_i],network.model.passive_branch_p[branch[0],branch[1],sn]),(1,network.model.passive_branch_s_nom[b[0],b[1]])],">=",0] for b in sub._extendable_branches.index for sn in snapshots})
+
+
+    l_constraint(network.model,"contingency_flow_upper",flow_upper,branch_outage_keys,snapshots)
+
+    l_constraint(network.model,"contingency_flow_lower",flow_lower,branch_outage_keys,snapshots)
+
+
+def add_contingency_constraints_lowmem(network, snapshots):
+
+    n = network
+
+    if not hasattr(n, "_branch_outages"):
+        n._branch_outages = n.passive_branches().index
+
+    branch_outages = [b
+        if isinstance(b, tuple) else ("Line", b)
+        for b in n._branch_outages
+    ]
+
+    comps = n.passive_branch_components & set(n.variables.index.levels[0])
+    if len(comps) == 0: return
+    dispatch_vars = pd.concat({c: get_var(n, c, "s") for c in comps}, axis=1)
+
+    invest_vars = pd.concat({
+        c: get_var(n, c, "s_nom")
+        if not get_extendable_i(n, c).empty
+        else pd.Series(dtype=np.float64)
+        for c in comps
+    })
+
+    constraints = {}
+    for sn in n.sub_networks.obj:
+
+        sn.calculate_BODF()
+
+        branches_i = sn.branches_i()
+        branches = sn.branches()
+        outages = branches_i.intersection(branch_outages)
+        
+        ext_i = branches.loc[branches.s_nom_extendable].index
+        fix_i = branches.loc[~branches.s_nom_extendable].index
+
+        p = dispatch_vars.loc[:, branches_i]
+        
+        BODF = pd.DataFrame(sn.BODF, index=branches_i, columns=branches_i)
+
+        lhs = {}
+        rhs = {}
+
+        for outage in outages:
+
+            def contingency_flow(branch):
+                if branch.name == outage: return ""
+                flow = linexpr((1, branch))
+                added_flow = linexpr((BODF.loc[branch.name, outage], p.loc[:, outage]))
```

Comment:
```suggestion
                added_flow = linexpr((BODF.at[branch.name, outage], p.loc[:, outage]))
```
`at` is faster for single element indexing